launch_testing/launch_testing/pytest/fixture.py

```diff
@@ -0,0 +1,88 @@
+# Copyright 2021 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import asyncio
+import inspect
+
+import launch
+import pytest
+
+try:
+    from _pytest.scope import Scope
+
+    def scope_gt(scope1, scope2):
+        return Scope(scope1) > Scope(scope2)
+except ImportError:
+    from _pytest.fixtures import scopemismatch as scope_gt
+
+
+def get_launch_service_fixture(scope='function'):
+
+    @pytest.fixture(scope=scope)
+    def launch_service():
+        ls = launch.LaunchService()
+        yield ls
+        assert ls._is_idle(), (
+            'launch service must be shut down before fixture tear down'
+        )
+    launch_service._launch_testing_overridable_fixture = True
+    return launch_service
+
+
+def get_event_loop_fixture(scope='function'):
+
+    @pytest.fixture(scope=scope)
+    def event_loop():
+        loop = asyncio.get_event_loop_policy().new_event_loop()
+        yield loop
+        loop.close()
+    event_loop._launch_testing_overridable_fixture = True
+    event_loop._launch_testing_fixture_scope = scope
+    return event_loop
+
+
+def fixture(*args, **kwargs):
+    """
+    Decorate launch_test fixtures.
+
+    For documentation on the supported arguments, see
+    https://docs.pytest.org/en/latest/reference/reference.html#pytest-fixture.
+    """
+    # Automagically override the event_loop and launch_testing fixtures
+    # with a fixture of the correct scope.
+    # This is not done if the user explicitly provided this fixture,
+    # they might get an ScopeError if not correctly defined.
+    scope = kwargs.get('scope', 'function')
+    if scope != 'function':
+        frame = inspect.stack()[1]
+        mod = inspect.getmodule(frame[0])
+        mod_locals = vars(mod)
+        for name, getter in (
+            ('launch_service', get_launch_service_fixture),
+            ('event_loop', get_event_loop_fixture)
+        ):
+            if name in mod_locals:
+                obj = mod_locals[name]
+                if (
+                    getattr(obj, '_launch_testing_overridable_fixture', False) and
+                    scope_gt(scope, obj._launch_testing_fixture_scope)
+                ):
+                    mod_locals[name] = getter(scope)
+            else:
+                mod_locals[name] = getter(scope)
```

Comment:
@ivanpauno I think I understand the override-able fixture magic, but doesn't this imply fixtures of different scopes cannot coexist in the same source file? I mean, you could have 3 tests using a module scope fixture and 3 tests using a function scope fixture and there wouldn't be any overlap, would it?

Reply:
If you have one launch_test fixture with `class` scope and one with `function` scope, the launch_service function will be class scoped.

This is required because of the following issue:

```python3

@pytest.fixture(scope='module')
def foo():
  ...

@pytest.fixture()
def bar(foo):
  ...
```

In that case you get a "scope mismatch error", because you cannot access a fixture with a lower scope from another fixture.

The fact that the launch_service/event_loop is automatically redefined with a larger scope only means that it might be shared by both launch_test fixtures, which is not necessarly a problem.
We could actually create one launch_service per launch_test fixture, and don't use a fixture for that at all.
For the event_loop though, you can only have one active event loop in a thread, so it would be an issue anyway.

Maybe in the future pytest solves this issue, see https://github.com/pytest-dev/pytest/issues/6101.

Reply:
> In that case you get a "scope mismatch error", because you cannot access a fixture with a lower scope from another fixture.

Aye, the `launch_service` fixture scope has to be at least as large as the largest scope of all launch fixtures found (BTW isn't that example reversed?).

> The fact that the launch_service/event_loop is automatically redefined with a larger scope only means that it might be shared by both launch_test fixtures, which is not necessarly a problem.

I see two problems.  Unrelated larger scope fixtures creeping into tests with smaller scope fixtures, and a large scope launch fixture being shutdown (which IIRC is a terminal state) by a smaller scope fixture upon finalization. The first one is not really a problem if we enforce a 1-to-1 relationship between tests and launch fixtures. The second one is.

> We could actually create one launch_service per launch_test fixture, and don't use a fixture for that at all.

Having separate instances is probably easiest, yeah. Alternatively, we can make `LaunchService` shutdowns not a terminal state (i.e. allow re-running after a shutdown).

Reply:
> (BTW isn't that example reversed?).

lol yes, fixed.

> and a large scope launch fixture being shutdown (which IIRC is a terminal state) by a smaller scope fixture upon finalization

Yes, I think we can detect that case and warn the user though.